snakemake/parser.py

```diff
@@ -103,11 +107,33 @@ def indentation(self, token):
             self.indent = token.end[1] - self.base_indent
             self.was_indented |= self.indent > 0
 
+    def parse_fstring(self, token: tokenize.TokenInfo):
+        isin_fstring = 1
+        t = token.string
+        for t1 in self.snakefile:
+            if t1.type == tokenize.FSTRING_START:
+                isin_fstring += 1
+                t += t1.string
+            elif t1.type == tokenize.FSTRING_END:
+                isin_fstring -= 1
+                t += t1.string
+            elif t1.type == tokenize.FSTRING_MIDDLE:
+                t += t1.string.replace("{", "{{").replace("}", "}}")
+            else:
+                t += t1.string
+            if isin_fstring == 0:
+                break
+        if hasattr(self, "cmd") and self.cmd[-1][1] == token:
+            self.cmd[-1] = t, token
```

Comment:
Can you please add a comment why this works for both >=3.12 and <3.12?

Reply:
Line 113-127 will never be touched in python<3.12, because in L135, it will test `is_fstring(token)`, which `return sys.version_info >= (3, 12) and token.type == tokenize.FSTRING_START`

Reply:
Thanks!